mapeditor/inspector/townbuildingswidget.cpp

```diff
@@ -68,6 +69,56 @@ std::string defaultBuildingIdConversion(BuildingID bId)
 	}
 }
 
+QStandardItem * getBuildingParentFromTreeModel(const CBuilding * building, QStandardItemModel & model)
+{
+	QStandardItem * parent = nullptr;
+	std::vector<QModelIndex> stack(1);
+	while (!parent && !stack.empty())
+	{
+		auto pindex = stack.back();
+		stack.pop_back();
+		auto rowCount = model.rowCount(pindex);
+		for (int i = 0; i < rowCount; ++i)
+		{
+			QModelIndex index = model.index(i, 0, pindex);
+			if (building->upgrade.getNum() == model.itemFromIndex(index)->data(MapEditorRoles::BuildingIDRole).toInt())
+			{
+				parent = model.itemFromIndex(index);
+				break;
+			}
+			if (model.hasChildren(index))
+				stack.push_back(index);
+		}
+	}
+	return parent;
+}
+
+std::list<QVariant> getBuildingVariantsFromModel(QStandardItemModel & model, int modelColumn, Qt::CheckState checkState)
+{
+	std::list<QVariant> result;
```

Comment:
list is just for easier creation of QVariantList, and QVariant is because of the same reason. QVariantList is then saved into QVariantMap that is stored as townEventListItem data

Reply:
- QVariantList is **not** `std::list` under the hood
- you can create it from any container that has begin/end

use vector if you don't need special behavior of some other container

Reply:
I know it's not std::list but it have convenient api for conversion from it. I had problems with using set (and I think the same with vector) for construction of QVaraintList. I already cleaned git history but I may double check that in the evening

Reply:
this should work:
```cpp
std::vector<QVariant> vec;
QVariantList list{vec.cbegin(), vec.cend()};
```

Reply:
even before qt 5.14?

Reply:
Why not simply use QVariantList? Mixing std and Qt containers is a bit weird approach IMO

Reply:
Perhaps that will be better solution, will check it